### ISSUE TITLE
Compilation Database

### ISSUE THREAD

**Issue by ghost** [PacktPublishing/Hands-On-Functional-Programming-with-Cpp#1]
Hi Alex,

just a quick question: How did you create the compilation database - `compile_commands.json` - and how can I possibly use it?

Kind regards,
Hans.

**Comment by alexboly** [comment]
Hi Hans,

Thank you for your question. The short answer is: the compilation database helps with the semantic auto-complete in vim. For details, see below.

To write the code, I used vim because it helps me achieve focus, it's very lightweight, and I can easily run the tests without leaving the editor. But I needed auto-complete support - because I'm lazy.

I finally found the brilliant YouCompleteMe plugin - https://github.com/ycm-core/YouCompleteMe. It does a great job, but it needs a bit of help for semantic auto-complete - for example, finding the functions in an STL container. That's where the compilation database falls into place.

As to how it's created, I just used a special tool called bear - https://github.com/rizsotto/Bear in combination with make. Whenever starting to work on a new file, I created the file, then the makefile, then run "bear make" and that's it - I have semantic auto-complete in vim!

Hope this helps,
Alex
